### PR TITLE
Feature/results reports update

### PR DESCRIPTION
**Pull request by yurii-krupa**
Fixed PR #38

### CODE REVIEW

**Comment by yurii-krupa** [comment]
[Results](https://trello.com/c/fOT1dZjp/25-results)

File: DTAdmin.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1,940 +1,996 @@
...
+        3F2BF59E1F9F61F600B2FA7E /* Subjects.storyboard in Resources */ = {isa = PBXBuildFile; fileRef = 3F2BF5981F9F61F600B2FA7E /* Subjects.storyboard */; };
+        3F5C2CE71FBB053600C05E73 /* SubjectTest.swift in Sources */ = {isa = PBXBuildFile; fileRef = 3F5C2CE61FBB053600C05E73 /* SubjectTest.swift */; };
+        3F5C2CE91FBB056100C05E73 /* QuestionTest.swift in Sources */ = {isa = PBXBuildFile; fileRef = 3F5C2CE81FBB056100C05E73 /* QuestionTest.swift */; };
+        3F5C2CEB1FBB05A200C05E73 /* AnswerTest.swift in Sources */ = {isa = PBXBuildFile; fileRef = 3F5C2CEA1FBB05A200C05E73 /* AnswerTest.swift */; };
+        3F94FD5E1FBC7C8C004F2F74 /* TestResults.swift in Sources */ = {isa = PBXBuildFile; fileRef = 3F94FD5D1FBC7C8C004F2F74 /* TestResults.swift */; };
+        3F94FD621FBC7C9F004F2F74 /* ResultByGroupViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 3F94FD5F1FBC7C9E004F2F74 /* ResultByGroupViewController.swift */; };
+        3F94FD631FBC7C9F004F2F74 /* Result.storyboard in Resources */ = {isa = PBXBuildFile; fileRef = 3F94FD601FBC7C9E004F2F74 /* Result.storyboard */; };
```
Comment (iostaf) [inline DTAdmin.xcodeproj/project.pbxproj]: Please fix this indentation mess.

File: DTAdmin/AppDelegate.swift [hunk 2]
```diff
@@ -89,5 +89,14 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         }
     }
 
+  // MARK: Custom Methods
+  class func getAppDelegate() -> AppDelegate {
+    return UIApplication.shared.delegate as! AppDelegate
```
Comment (iostaf) [inline DTAdmin/AppDelegate.swift]: We use 4 spaces as an indentation on this project. Please stick to the common rules.

File: DTAdmin/Models/DataManager.swift [hunk 3]
```diff
@@ -583,6 +583,153 @@ class DataManager: HTTPManager {
             completionHandler(recordList, nil)
         }
     }
+    
+    func getResultsBy(group: [String: String], subject: [String: String], test: [String: String], maxMark: String, completionHandler: @escaping (_ error: String?, _ students: [TestResults]?) -> ()) {
```
Comment (iostaf) [inline DTAdmin/Models/DataManager.swift]: It's worth documenting all these parameters.

File: DTAdmin/Models/DataManager.swift [hunk 4]
```diff
@@ -583,6 +583,153 @@ class DataManager: HTTPManager {
...
+        
+        guard let cookie = StoreHelper.getCookie() else { return }
+        let httpManager = HTTPManager()
+        guard let url = URL(string: httpManager.urlProtocol + httpManager.urlDomain + "/Result/GetRecordsByTestGroupDate/" + test["id"]! + "/" + group["id"]!) else { return }
+        
+        var request = URLRequest(url: url)
+        request.httpMethod = "GET"
+        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+        request.setValue("utf-8", forHTTPHeaderField: "Charset")
+        request.setValue(cookie[Keys.cookie], forHTTPHeaderField: Keys.cookie)
```
Comment (iostaf) [inline DTAdmin/Models/DataManager.swift]: Why do not use common DataManager here?

File: DTAdmin/Models/DataManager.swift [hunk 5]
```diff
@@ -583,6 +583,153 @@ class DataManager: HTTPManager {
...
+        
+        guard let cookie = StoreHelper.getCookie() else { return }
+        let httpManager = HTTPManager()
+        guard let url = URL(string: httpManager.urlProtocol + httpManager.urlDomain + "/Result/getResultTestIdsByGroup/" + id) else { return }
+        
+        var request = URLRequest(url: url)
+        request.httpMethod = "GET"
+        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+        request.setValue("utf-8", forHTTPHeaderField: "Charset")
+        request.setValue(cookie[Keys.cookie], forHTTPHeaderField: Keys.cookie)
```
Comment (iostaf) [inline DTAdmin/Models/DataManager.swift]: Why do not use common DataManager here?

File: DTAdmin/Models/DataManager.swift [hunk 6]
```diff
@@ -583,6 +583,153 @@ class DataManager: HTTPManager {
...
+        
+        guard let cookie = StoreHelper.getCookie() else { return }
+        let httpManager = HTTPManager()
+        guard let url = URL(string: httpManager.urlProtocol + httpManager.urlDomain + "/EntityManager/getEntityValues") else { return }
+        
+        var request = URLRequest(url: url)
+        request.httpMethod = "POST"
+        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+        request.setValue("utf-8", forHTTPHeaderField: "Charset")
+        request.setValue(cookie[Keys.cookie], forHTTPHeaderField: Keys.cookie)
```
Comment (iostaf) [inline DTAdmin/Models/DataManager.swift]: Why do not use common DataManager here?

File: DTAdmin/Models/DataManager.swift [hunk 7]
```diff
@@ -583,6 +583,153 @@ class DataManager: HTTPManager {
...
+    func getSubjectsBy(ids: [String], completionHandler: @escaping (_ error: String?, _ subjects: [SubjectStructure]?) -> ()) {
+        
+        guard let cookie = StoreHelper.getCookie() else { return }
+        let httpManager = HTTPManager()
+        guard let url = URL(string: httpManager.urlProtocol + httpManager.urlDomain + "/EntityManager/getEntityValues") else { return }
+        var request = URLRequest(url: url)
+        request.httpMethod = "POST"
+        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+        request.setValue("utf-8", forHTTPHeaderField: "Charset")
+        request.setValue(cookie[Keys.cookie], forHTTPHeaderField: Keys.cookie)
```
Comment (iostaf) [inline DTAdmin/Models/DataManager.swift]: Why do not use common DataManager here?

File: DTAdmin/Printer/PreviewViewController.swift [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        //Unwrap all nessesary data
+        guard let group = self.quizParameters["Group"],
+            let groupId = group["id"],
+            let subject = self.quizParameters["Subject"],
+            let test = self.quizParameters["Quiz"],
+            let testId = test["id"] else {
+                self.showWarningMsg("Quiz has wrong parameters please try again")
+                return
+        }
+        // Get sudents test passed by group
```
Comment (iostaf) [inline DTAdmin/Printer/PreviewViewController.swift]: Typo in `sudents`

File: DTAdmin/Printer/PreviewViewController.swift [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    }
+    
+    @objc private func shareReport() {
+        pdfComposer.exportHTMLContentToPDF(HTMLContent: HTMLContent)
+        self.showOptionsAlert()
+    }
+    
+    func showOptionsAlert() {
+        let alertController =
+            UIAlertController(title: NSLocalizedString("Yeah!", comment: "Print alert title"),
```
Comment (iostaf) [inline DTAdmin/Printer/PreviewViewController.swift]: Yeah! is funny but not user-friendly

File: DTAdmin/Printer/PreviewViewController.swift [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+                let request = URLRequest(url: url)
+                self.webPreview.loadRequest(request)
+            }
+        }
+        
+        let actionEmail = UIAlertAction(title: NSLocalizedString("Send via Email", comment: "Send email UIAlert option"),
+                                        style: UIAlertActionStyle.default) { (action) in
+            DispatchQueue.main.async { self.sendEmail() }
+        }
+        let actionNothing = UIAlertAction(title: NSLocalizedString("Nothing", comment: "Do nothing UIAlert option"),
```
Comment (iostaf) [inline DTAdmin/Printer/PreviewViewController.swift]: Why do we need this *Nothing* option?

File: DTAdmin/Printer/PrintPageRenderer.swift [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    func getTextSize(text: String, font: UIFont!, textAttributes: [NSAttributedStringKey: AnyObject]! = nil) -> CGSize {
+        let testLabel = UILabel(frame: CGRect(x: 0.0, y: 0.0, width: self.paperRect.size.width, height: footerHeight))
+        if let attributes = textAttributes {
+            testLabel.attributedText = NSAttributedString(string: text, attributes: attributes)
+        }
+        else {
+            testLabel.text = text
+            testLabel.font = font!
```
Comment (iostaf) [inline DTAdmin/Printer/PrintPageRenderer.swift]: Please do not use force unwrapping.

File: DTAdmin/Result/ResultByGroupViewController.swift [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+        var subjectsIds = [String]()
+        
+        DataManager.shared.getResultTestIds(byGroup: groupId) { (error, testIds) in
+            if let error = error {
+                self.stopActivity()
+                self.showWarningMsg(error)
+            } else {
+                guard let testIds = testIds else {
+                    self.stopActivity()
+                    self.showWarningMsg("Error occured during data handle")
```
Comment (iostaf) [inline DTAdmin/Result/ResultByGroupViewController.swift]: Please use NSLocalizedString for all user faced messages.

File: DTAdmin/Result/ResultByGroupViewController.swift [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+                    testsIds.append(i["test_id"]!)
+                }
+                
+                DataManager.shared.getTestsBy(ids: testsIds, completionHandler: { (error, tests) in
+                    if let error = error {
+                        self.stopActivity()
+                        self.showWarningMsg(error)
+                    } else {
+                        guard var tests = tests else {
+                            self.showWarningMsg("Wrong API response")
```
Comment (iostaf) [inline DTAdmin/Result/ResultByGroupViewController.swift]: Please use NSLocalizedString for all user faced messages.

File: DTAdmin/Result/ResultByGroupViewController.swift [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+    }
+    
+}
+
+extension ResultByGroupViewController: UITableViewDelegate {
+}
+
+extension ResultByGroupViewController: UITableViewDataSource {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return tests != nil ? tests!.count : 0
```
Comment (iostaf) [inline DTAdmin/Result/ResultByGroupViewController.swift]: Please do not use force unwrapping.

**Review by iostaf — CHANGES_REQUESTED** [review]

File: DTAdmin/Printer/PreviewViewController.swift (hunk 10 above)
Comment (yurii-krupa) [reply DTAdmin/Printer/PreviewViewController.swift]: Replace to close option?

File: DTAdmin/Result/ResultByGroupViewController.swift (hunk 14 above)
Comment (yurii-krupa) [reply DTAdmin/Result/ResultByGroupViewController.swift]: So should I rewrite it with "if let" statement?

**Review by iostaf — DISMISSED** [review]

**Pull request closed without merging**